inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/AsyncClassTransformer.java

```diff
@@ -130,8 +129,19 @@ private void deinstrumentAllClasses() {
             try {
                 instrumentation.retransformClasses(batchClasses.toArray(new Class[]{}));
                 Thread.yield(); //yield to allow the target application to do stuff between the batches
-            } catch (UnmodifiableClassException e) {
-                log.error("Error removing applied transformations", e);
+            } catch (Exception e) {
+                if (batchClasses.size() > 1) {
+                    log.error("Error removing applied transformations for batch, retrying classes one by one", e);
+                    for (Class<?> clazz : batchClasses) {
+                        try {
+                            instrumentation.retransformClasses(clazz);
+                        } catch (Exception e2) {
+                            log.error("Error removing applied transformations of class '{}'", clazz.getName(), e2);
+                        }
+                    }
+                } else {
```

Comment:
this will throw NPE if the batchClasses is empty, at least by looking into this part of the code it can happen..